### ISSUE TITLE
How to get the last inserted rowid

### ISSUE THREAD

**Issue by acautin** [oracle/odpi#111]
I am trying to implement the following code using odpi:

https://github.com/K2InformaticsGmbH/erloci/blob/885269045ea622ad448afe199fcdd45170950ac4/c_src/erloci_lib/ocistmt.cpp#L291-L328

But I am missing how to ask for the rowid attributes for the newly inserted rows, is there a function I can call to get this information ?

**Comment by cjbj** [comment]
Oracle DB doesn't have this capability unless you add a RETURNING INTO clause to the statement.  You can return the ROWID or an auto-generated key, such as from a sequence in older DBs, or from an identity column in newer DBs.
```
INSERT INTO tab VALUES (:1) RETURNING ROWID INTO :2
```

**Comment by c-bik** [comment]
@cjbj (from @acautin link above) in erloci we could retrive Row ID from from any DML implicitely as follows.
```c
OCIRowid *pRowID;
OraText *rowID;
OCIDescriptorAlloc(envhp, &pRowID, OCI_DTYPE_ROWID, 0, NULL);
OCIAttrGet(stmthp, OCI_HTYPE_STMT, pRowID, 0, OCI_ATTR_ROWID, errhp);
OCIRowidToChar(pRowID, rowID, &size, errhp);
```
This didn't  require any `RETURNING INTO` in original SQL.

Is there something equivalent to this possible with dpi APIs?

**Comment by c-bik** [comment]
@cjbj I quickly put together the following to get the ROWID through `RETURNING`...
```c
    dpiContext *context, dpiErrorInfo errorInfo;
    dpiConn *conn;
    dpiVar *var, dpiData *data;
    dpiStmt *stmt;
    char * rowidStr[100];

    dpiContext_create(DPI_MAJOR_VERSION, DPI_MINOR_VERSION, &context, &errorInfo);
    dpiConn_create(context, ..., &conn);
    const char *Sql = "insert into test (col2) values ('efg') returning rowid into :rid";
    dpiConn_prepareStmt(conn, ..., Sql, ..., &stmt);
    dpiConn_newVar(conn, DPI_ORACLE_TYPE_ROWID, DPI_NATIVE_TYPE_ROWID, 1, 100, true, false, null, &var, &data);
    dpiStmt_bindByName(stmt, "rid", 3, var);
    dpiStmt_execute(stmt, DPI_MODE_EXEC_DEFAULT, ...);
    dpiRowid_getStringValue(data->value.asRowid, &rowidStr, ...);
    printf("rowid %s\r\n", rowidStr);
```
the `rowidStr` is empty while None of of the dpi calls returned any error!

what am I missing?

TIA

**Comment by acautin** [comment]
@cjbj any update on how to use something similar to `OCIAttrGet` with odpi or pointers on how to call it directly ?

**Comment by kubo** [comment]
@acautin 
> any update on how to use something similar to OCIAttrGet with odpi or pointers on how to call it directly? 

How about using internal functions declared in [dpiImpl.h](https://github.com/oracle/odpi/blob/master/src/dpiImpl.h)? Note that they may change between minor updates.
```c
// I have not tested this code. It may not pass compilation.
#include <dpiImpl.h>
#define OCI_ATTR_ROWID   19

int dpiStmt_getRowid(dpiStmt *stmt, dpiRowid **rowid)
{
    dpiError error;
    int status;

    if (dpiStmt__check(stmt, __func__, &error) < 0)
        return dpiGen__endPublicFn(stmt, DPI_FAILURE, &error);
    if (dpiRowid__allocate(stmt->conn, rowid, error) < 0)
        return dpiGen__endPublicFn(stmt, DPI_FAILURE, &error);
    status = dpiOci__attrGet(stmt->handle, DPI_OCI_HTYPE_STMT, (*rowid)->handle,
            0, OCI_ATTR_ROWID, "get rowid", &error);
    return dpiGen__endPublicFn(stmt, status, &error);
}

...
   dpiRowid *rowid;
   const char *value;
   uint32_t valuelen;
   dpiStmt_getRowid(stmt, &rowid);
   dpiRowid_getStringValue(rowid, &value, &valuelen);
   ... use value and valuelen ...
   dpiRowid_release(rowid);
```

@c-bik 
> the rowidStr is empty while None of of the dpi calls returned any error!
>
> what am I missing?

[dpiVar_getReturnedData](https://oracle.github.io/odpi/doc/functions/dpiVar.html#c.dpiVar_getReturnedData) ([sample code](https://github.com/oracle/odpi/blob/v3.2.1/samples/DemoDMLReturning.c#L89-L101))

**Comment by c-bik** [comment]
@kubo 👍

**Comment by acautin** [comment]
Thanks for the support I think we got enough information to continue.

**Comment by cjbj** [comment]
Let's leave this open so we can decide if any ODPI-C changes are needed. 
@anthony-tuininga has just got back and may have some comments.

**Comment by c-bik** [comment]
@cjbj Thanks. The `RETURNING INTO` worked and we are now have `dpiVar_getReturnedData` in oranif. In DDErl @acautin is going with that unless/until anything similar to `dpiStmt_getRowid` is officially supported in odpi.

**Comment by anthony-tuininga** [comment]
@c-bik, as already mentioned, using ``dpiVar_getReturnedData()`` is the correct way to do this. This call is needed since a particular statement may return more than one "row" of data in the DML returning clause. The [sample](https://github.com/oracle/odpi/blob/master/samples/DemoDMLReturning.c) provided for DML returning demonstrates that fact. In the plain insert case that isn't strictly necessary since only one "row" is returned, but I require the call to ``dpiVar_getReturnedData()`` in order to be consistent in all cases.

Adding support for a call to get the "last inserted rowid" is unlikely to ever take place unless and until the Oracle Database provides that information. Without the help of the database the SQL statement would need to be modified to append the RETURNING clause and a new variable bound -- and that could potentially be an issue.

**Comment by c-bik** [comment]
> Without the help of the database the SQL statement would need to be modified to append the RETURNING clause and a new variable bound -- and that could potentially be an issue.

I suppose this was also the reason which motivated @acautin to open this issue.

**Comment by anthony-tuininga** [comment]
@c-bik, I just read through this again and noted that there is an additional capability mentioned by @acautin -- but that appeared to only be valid for positioned updates. I'll do some experimenting and see what I can come up with.

**Comment by anthony-tuininga** [comment]
So it looks like that approach works for regular DML statements as well. So I can create a function that looks like this:

```
int dpiStmt_getLastRowid(dpiStmt *stmt, dpiRowid **rowid);
```

The code that manages this function looks similar to what @kubo suggested. I think, however, that rather than force the rowid to be allocated and destroyed for each call, that the first call to this would allocate a rowid and the reference would be stored on the statement (and destroyed when the statement itself is destroyed). Subsequent calls would simply return this reference directly. That way you don't have to continuously destroy the rowid after each call. What do you think of that approach?

**Comment by c-bik** [comment]
@anthony-tuininga 
> That way you don't have to continuously destroy the rowid after each call. What do you think of that approach?

Perfect!

**Comment by cjbj** [comment]
@anthony-tuininga if you're going to explicitly call out `dpiRowid_addRef()`, you may as well mention the inverse `dpiRowid_release()`?

**Comment by anthony-tuininga** [comment]
This was released with ODPI-C 3.3.
